File: app/src/main/java/com/stevenschoen/putionew/PutioOpenFileService.java
package com.stevenschoen.putionew;

import android.app.DownloadManager;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.database.Cursor;
import android.os.IBinder;

import com.stevenschoen.putionew.activities.FileFinished;

public class PutioOpenFileService extends Service {
	DownloadManager downloadManager;
	long[] downloadIds;
	int id;
	String filename;
	int mode;
	
	IntentFilter fileDownloadUpdateIntentFilter = new IntentFilter(PutioActivity.fileDownloadUpdateIntent);

	@Override
	public void onCreate() {
		super.onCreate();

		downloadManager = (DownloadManager) getSystemService(DOWNLOAD_SERVICE);
	}

	@Override
	public int onStartCommand(Intent intent, int flags, int startId) {
		this.downloadIds = intent.getExtras().getLongArray("downloadIds");
		this.id = intent.getExtras().getInt("id");
		this.filename = intent.getExtras().getString("filename");
		this.mode = intent.getExtras().getInt("mode");
		
		IntentFilter intentFilter = new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE);
		registerReceiver(downloadReceiver, intentFilter);
		// We want this service to continue running until it is explicitly
		// stopped, so return sticky.
		return Service.START_STICKY;
	}
	
	private BroadcastReceiver downloadReceiver = new BroadcastReceiver() {
		@Override
		public void onReceive(Context arg0, Intent arg1) {
            if (mode != PutioUtils.ACTION_NOTHING) {
                for (long downloadId : downloadIds) {
                    DownloadManager.Query query = new DownloadManager.Query();
                    query.setFilterById(downloadId);
                    Cursor cursor = downloadManager.query(query);

                    if (cursor.moveToFirst()) {
                        int columnIndex = cursor
                                .getColumnIndex(DownloadManager.COLUMN_STATUS);
                        int status = cursor.getInt(columnIndex);
                        int columnReason = cursor
                                .getColumnIndex(DownloadManager.COLUMN_REASON);
                        int reason = cursor.getInt(columnReason);

                        switch (status) {
                            case DownloadManager.STATUS_SUCCESSFUL:
                                Intent finishedIntent = new Intent(PutioOpenFileService.this, FileFinished.class);
                                finishedIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                                finishedIntent.putExtra("downloadId", downloadId);
                                finishedIntent.putExtra("id", id);
                                finishedIntent.putExtra("filename", filename);
                                finishedIntent.putExtra("mode", mode);
                                startActivity(finishedIntent);
                                break;
                            case DownloadManager.STATUS_FAILED:
                                break;
                            case DownloadManager.STATUS_PAUSED:
                                break;
                            case DownloadManager.STATUS_PENDING:
                                break;
                            case DownloadManager.STATUS_RUNNING:
                                break;
                        }
                    }
                }
            }
		}

	};

	@Override
	public void onDestroy() {
		unregisterReceiver(downloadReceiver);
	}

	@Override
	public IBinder onBind(Intent arg0) {
		return null;
	}
}

File: app/src/main/java/com/stevenschoen/putionew/PutioOpenFileService.kt
package com.stevenschoen.putionew

import android.app.DownloadManager
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.Uri
import android.os.IBinder
import com.stevenschoen.putionew.files.FileFinishedActivity

class PutioOpenFileService : Service() {

    companion object {
        const val EXTRA_DOWNLOAD_ID = "download_id"
    }

    lateinit var downloadManager: DownloadManager

    var downloadId: Long = -1

    override fun onCreate() {
        super.onCreate()

        downloadManager = getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        super.onStartCommand(intent, flags, startId)

        this.downloadId = intent.extras.getLong(EXTRA_DOWNLOAD_ID, -1)

        val intentFilter = IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE)
        registerReceiver(downloadReceiver, intentFilter)

        return Service.START_STICKY
    }

    private val downloadReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val query = DownloadManager.Query().setFilterById(downloadId)
            val cursor = downloadManager.query(query)

            if (cursor.moveToFirst()) {
                val status = cursor.getInt(cursor.getColumnIndex(DownloadManager.COLUMN_STATUS))
                when (status) {
                    DownloadManager.STATUS_SUCCESSFUL -> {
                        val title = cursor.getString(cursor.getColumnIndex(DownloadManager.COLUMN_TITLE))
                        val uri = Uri.parse(cursor.getString(cursor.getColumnIndex(DownloadManager.COLUMN_LOCAL_URI)))
                        val type = cursor.getString(cursor.getColumnIndex(DownloadManager.COLUMN_MEDIA_TYPE))

                        val finishedIntent = Intent(this@PutioOpenFileService, FileFinishedActivity::class.java).apply {
                            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                            putExtra(FileFinishedActivity.EXTRA_NAME, title)
                            putExtra(FileFinishedActivity.EXTRA_URI, uri)
                            putExtra(FileFinishedActivity.EXTRA_MEDIA_TYPE, type)
                        }
                        startActivity(finishedIntent)

                        stopSelf()
                    }
                    DownloadManager.STATUS_FAILED -> { }
                    DownloadManager.STATUS_PAUSED -> { }
                    DownloadManager.STATUS_PENDING -> { }
                    DownloadManager.STATUS_RUNNING -> { }
                }
            }
        }
    }

    override fun onDestroy() {
        unregisterReceiver(downloadReceiver)
    }

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }
}